src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp

```diff
@@ -0,0 +1,193 @@
+                                maxIterations,
+                                snapshots)))
+{
+  /* Nothing to do here */
+}
+
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType,
+    typename DecayPolicyType
+>
+template<typename PolicyType>
+SGDR<
+    DecomposableFunctionType,
+    UpdatePolicyType,
+    DecayPolicyType
+>::SGDR(DecomposableFunctionType& function,
+        const size_t epochRestart,
+        const double multFactor,
+        const size_t batchSize,
+        const double stepSize,
+        const size_t maxIterations,
+        const double tolerance,
+        const bool shuffle,
+        const UpdatePolicyType& updatePolicy,
+        const typename std::enable_if_t<std::is_same<
+            PolicyType, CyclicalDecay>::value>* /* junk */) :
+    function(function),
+    batchSize(batchSize),
+    accumulate(true),
+    optimizer(OptimizerType(function,
+                            batchSize,
+                            stepSize,
+                            maxIterations,
+                            tolerance,
+                            shuffle,
+                            updatePolicy,
+                            CyclicalDecay(
+                                epochRestart,
+                                multFactor,
+                                stepSize,
+                                batchSize,
+                                function.NumFunctions())))
+{
+  /* Nothing to do here */
+}
+
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType,
+    typename DecayPolicyType
+>
+template<typename PolicyType>
+double SGDR<
+    DecomposableFunctionType,
+    UpdatePolicyType,
+    DecayPolicyType
+>::Optimize(arma::mat& iterate,
+            const typename std::enable_if_t<std::is_same<
+                PolicyType, SnapshotEnsembles>::value>* /* junk */)
+{
+  // If a user changed the step size he hasn't update the step size of the
+  // cyclical decay instantiation, so we have to do here.
+  if (optimizer.StepSize() != optimizer.DecayPolicy().StepSize())
+  {
+    optimizer.DecayPolicy().StepSize() = optimizer.StepSize();
+  }
+
+  // If a user changed the batch size we have to update the restart fraction
+  // of the cyclical decay instantiation.
+  if (optimizer.BatchSize() != batchSize)
+  {
+    batchSize = optimizer.BatchSize();
+    optimizer.DecayPolicy().EpochBatches() = function.NumFunctions() /
+        double(batchSize);
+  }
+
+  double overallObjective = optimizer.Optimize(iterate);
+
+  // Accumulate snapshots.
+  if (accumulate)
+  {
+    for (size_t i = 0; i < optimizer.DecayPolicy().Snapshots().size(); ++i)
+    {
+      iterate += optimizer.DecayPolicy().Snapshots()[i];
+    }
+    iterate /= (optimizer.DecayPolicy().Snapshots().size() + 1);
+
+    // Calculate final objective.
+    overallObjective = 0;
+    for (size_t i = 0; i < function.NumFunctions(); ++i)
+      overallObjective += function.Evaluate(iterate, i);
+  }
```

Comment:
But this part I am less sure of how to move into `SnapshotEnsembles`.  The only thing I could think of is to make a template specialization for `SGD<..., SnapshotEnsembles>` and `MiniBatchSGD<..., SnapshotEnsembles>` but that would be ugly and weird because we can't do partial template member specialization.